forge/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -0,0 +1,230 @@
+    }
+}
+
+impl<DB> Inspector<DB> for Cheatcodes
+where
+    DB: Database,
+{
+    fn call(
+        &mut self,
+        data: &mut EVMData<'_, DB>,
+        call: &CallInputs,
+        _: bool,
+    ) -> (Return, Gas, Bytes) {
+        if call.contract == *CHEATCODE_ADDRESS {
+            match self.apply_cheatcode(data, call.context.caller, call) {
+                Ok(retdata) => (Return::Return, Gas::new(0), retdata),
+                Err(err) => (Return::Revert, Gas::new(0), err),
+            }
+        } else {
+            // Handle mocked calls
+            if let Some(mocks) = self.mocked_calls.get(&call.contract) {
+                if let Some(mock_retdata) = mocks.get(&call.input) {
+                    return (Return::Return, Gas::new(0), mock_retdata.clone())
+                } else if let Some((_, mock_retdata)) =
+                    mocks.iter().find(|(mock, _)| *mock == &call.input[..mock.len()])
+                {
+                    return (Return::Return, Gas::new(0), mock_retdata.clone())
+                }
+            }
+
+            (Return::Continue, Gas::new(0), Bytes::new())
+        }
+    }
+
+    fn initialize_interp(
+        &mut self,
+        interpreter: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        _: bool,
+    ) -> Return {
+        // Apply our prank
+        if let Some(prank) = &self.prank {
+            if data.subroutine.depth() >= prank.depth &&
+                interpreter.contract.caller == prank.prank_caller
+            {
+                // At the target depth we set `msg.sender`
+                if data.subroutine.depth() == prank.depth {
+                    interpreter.contract.caller = prank.new_caller;
+                }
+
+                // At the target depth, or deeper, we set `tx.origin`
+                if let Some(new_origin) = prank.new_origin {
+                    data.env.tx.caller = new_origin;
+                }
+            }
+        }
+
+        Return::Continue
+    }
+
+    fn step(&mut self, interpreter: &mut Interpreter, _: &mut EVMData<'_, DB>, _: bool) -> Return {
+        if let Some(storage_accesses) = &mut self.accesses {
+            match interpreter.contract.code[interpreter.program_counter()] {
+                opcode::SLOAD => {
+                    let key = try_or_continue!(interpreter.stack().peek(0));
+                    storage_accesses
+                        .reads
+                        .entry(interpreter.contract().address)
+                        .or_insert_with(Vec::new)
+                        .push(key);
+                }
+                opcode::SSTORE => {
+                    let key = try_or_continue!(interpreter.stack().peek(0));
+                    storage_accesses
+                        .writes
+                        .entry(interpreter.contract().address)
+                        .or_insert_with(Vec::new)
+                        .push(key);
+                }
```

Comment:
As mentioned in the PR, if we want to replicate the "SLOAD per SSTORE" behavior we just write to `storage_accesses.reads` here as well

Reply:
@brockelmore up to you here, wdyt?

Reply:
hmmm ideally we do, because it maintains simplicity for storage overriding based on function calls. otherwise in stdstorage, we would have to also check writes

Reply:
I'll replicate the behaviour in the port :)